### ISSUE TITLE
Alter value before validation

### ISSUE THREAD

**Issue by FrankFang** [handsontable/handsontable#3402]
Demo: http://jsfiddle.net/frankfang/ub5xe8nj/2/

Expected: 
1. Change birthday to 1990-01/01
2. 1990-01/01 turns to 1990-01-01 (See beforeChange hook)
3. Validator returns true

But what I got is:
1. Change birthday to 1990-01/01
2. 1990-01/01 turns to 1990-01-01 (See beforeChange hook)
3. Validator returns **false**

Help :)

**Comment by AMBudnik** [comment]
Hi @FrankFang 
Did you added correct demo? There's no `birthday` cell/row/col in this demo

**Comment by FrankFang** [comment]
Sorry, here it is http://jsfiddle.net/frankfang/ub5xe8nj/2/

**Comment by AMBudnik** [comment]
You're right.
I just added two `console.log` http://jsfiddle.net/handsoncode/wuc8g06n/ where we see that cell gets validated before it is changed and if we're using 'beforeChange' it should't happen.

**Comment by FrankFang** [comment]
Hi, any update?

**Comment by AMBudnik** [comment]
Hi @FrankFang 
There's one related issue at #3381, I have to discuss this with our team.

**Comment by rukayaj** [comment]
@FrankFang I'm needing to do the same thing as you. What I've done in the meantime while they sort it out is run my data reformatting/cleaning on beforeValidate AND beforeChange. So you have to do it twice which isn't ideal but it does work.

beforeValidate http://docs.handsontable.com/0.15.0-beta3/Hooks.html#event:beforeValidate
" used to manipulate value of changed cell before it is applied to the validator... this will not affect values of changes. This will change value ONLY for validation!"

This makes sure it validates correctly, and then beforeChange actually changes the value in the cell. Something like:
`
// My HoT options
var hotOptions;

// My regular expression to reformat time data
var timeRegex = /\s*hr?\s*|\s*:\s*/;

// Run to pass in the 'corrected' data to the validator, does not actually change value in cell
hotOptions['beforeValidate'] = function (value, row, prop, source) {
	if(prop == 3) { // 3 = column ID
		return value.replace(timeRegex, '.');
	}
};

// Run after the validation is done, changes value in the cell in the same way as on beforeValidate
hotOptions['beforeChange'] = function (changes, source) {
	for(var i = changes.length - 1; i >= 0; i--) {
	    if(changes[i][1] === 3) {
		  if(changes[i][3] !== parseInt(changes[i][3], 10)) {
			changes[i][3] = changes[i][3].replace(timeRegex, '.');
		  }
	    }
	}
};

// My column definitions
hotOptions['columns'] = [
      ....
      {
        type: 'numeric',
        format: '00',
        default: '00',
        validator: /^([01]?[0-9]|2[0-3])$/
        //validator: validateTime,
      },
      ...
];

// Instantiate hot instance with these hotOptions
`

**Comment by AMBudnik** [comment]
related: https://github.com/handsontable/handsontable/issues/3424

**Comment by AMBudnik** [comment]
http://jsfiddle.net/handsoncode/m9gsvx0e/ still present in 0.31.2

**Comment by AMBudnik** [comment]
In `7.4.0` if we

> Change birthday to 1990-01/01

We get `true` instead of `false` in the `afterValidate` hook. And (as it worked before) we get a call that a cell was changed, then validated. 

So there are **no changes since the last update.**

![image](https://user-images.githubusercontent.com/10757813/74139577-eafff300-4bf3-11ea-9874-d499bb151e1c.png)
https://codesandbox.io/s/demo-for-740-gh-3402-0dojl

**Comment by wojciechczerniak** [comment]
@AMBudnik I'm lost here. What is the expected result?

The original post was that the value is changed, but validated to `false`.
Now (v7 and v8) I see that value is changed and validated correctly (after the change).

The order: changed, then validated is IMO correct one. We couldn't change the value before validation if it wasn't.

**Comment by AMBudnik** [comment]
@wojciechczerniak 

I've tested the issue using `7.4.2` http://jsfiddle.net/gk56ubsf/ and yes, it is working as expected (returning `true`). To keep all the changes tested with beta2 I've also made a demo for 8.0.0-beta2-rev18 http://jsfiddle.net/528fx67t/ where we have the exact result.

@FrankFang @rukayaj you've waited long for this issue to get solved. I hope that it can still benefit you in your work. 

I'm closing this issue as solved.
